[SofaPython+CI] ISSofa merge: base branch

guparan opened the pull request:
This is the first PR of a long series concerning ISSofa (InSimo) commits merge.
This branch has to be merged first because it is the base branch of all further integrations.
It contains mostly SofaPython commits and also some CI and environment commits.

## issofa_python

### New features
Add binding python to get slaves and names on baseObjects.
Add Python bindings for triangle and point topology modifiers. Example scene which show triangle refining.
Creation of Binding_BaseTopologyObject common to PointSetTopologyModifier and TriangleSetTopologyModifier.
Add a getObjects() method to python BaseContext interface. Allow selection of objects based on type and name. May be extended to include more search options later.

### Improvements
SendScriptEvent should only be applied on a specified node, not on root !
WARNING: may affect python script behavior.
Allow to get the value of a data as a string when its type is not handle by the binding.
Object and type names are now both optional when calling BaseContext_getObjects(). Moreover the search direction can now optionally be passed to BaseContext_getObjects() (default is 'Local').
Binding_Data expose fullPath read only parameter for BaseData that derive from DataFileName, which as the name suggests, returns the fullPath to the file pointed by that data. Returns None otherwise.

### Bugfixes
If the data to be gotten using the bind is a vector of type 'unknown' then the text of the data is returned as a whole.
Fix compilation and warnings with gcc.
Fix SofaPython build on linux.
Fix undefined behaviour and potential crash in Python BaseContext.getObjects() due to PyList accessed out of bounds. 
Fix a memory leak which happens whenever an object is transferred from C++ to Python.



_______________________________________________________________
<!--- Please leave this at the end of your message -->
This PR: 
- [x] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings nor unit test failures.
- [x] does not break existing scenes.
- [x] does not break API compatibility.
- [x] has been reviewed and agreed to be transitional.
- [x] is more than 1 week old.

**Reviewers will merge only if all these checks are true.**

thomasjundinsimo committed:
ADD: binding python to get slaves and names on baseObjects

git-svn-id: https://code.insimo.fr/svn/IS/ISSofa/branches/ISSofa@10487 f8a8aa39-ebdb-4476-bae3-08f188f1b544

Conflicts:
	applications/plugins/SofaPython/Binding_BaseObject.cpp

thomasjundinsimo committed:
UPDATE: sendScriptEvent should only be applied on a specified node, not on root /!\ WARNING: may affect python script behavior

git-svn-id: https://code.insimo.fr/svn/IS/ISSofa/branches/ISSofa@10499 f8a8aa39-ebdb-4476-bae3-08f188f1b544

Conflicts:
	applications/plugins/SofaPython/Binding_Node.cpp

Commit:
CHANGE : if the data to be gotten using the bind is a vector of type 'unknown' then the text of the data is returned as a whole
MERGE CHANGE : include was already here, put amelioration : test line 137 on typeinfo->ValidInfo cleaner than test on name

git-svn-id: https://code.insimo.fr/svn/IS/ISSofa/branches/ISSofa@10519 f8a8aa39-ebdb-4476-bae3-08f188f1b544

Conflicts:
	applications/plugins/SofaPython/Binding_Data.cpp

Commit:
CHANGE : allow to get the value of a data as a string when its type is not handle by the binding

git-svn-id: https://code.insimo.fr/svn/IS/ISSofa/branches/ISSofa@10522 f8a8aa39-ebdb-4476-bae3-08f188f1b544

Conflicts:
	applications/plugins/SofaPython/Binding_Data.cpp

fjourdes committed:
ADD: Python bindings for triangle and point topology modifiers. Example scene which show triangle refining.

git-svn-id: https://code.insimo.fr/svn/IS/ISSofa/branches/ISSofa@10526 f8a8aa39-ebdb-4476-bae3-08f188f1b544

Conflicts:
	applications/plugins/SofaPython/Binding.cpp
	applications/plugins/SofaPython/CMakeLists.txt
	applications/plugins/SofaPython/PythonMacros.cpp

JeremieA committed:
FIX: compilation and warnings with gcc

git-svn-id: https://code.insimo.fr/svn/IS/ISSofa/branches/ISSofa@10527 f8a8aa39-ebdb-4476-bae3-08f188f1b544

neildavidforrest committed:
UPDATE: Add a getObjects() method to python BaseContext interface. Allow selection of objects based on type and name. May be extended to include more search options later.

git-svn-id: https://code.insimo.fr/svn/IS/ISSofa/branches/ISSofa@10540 f8a8aa39-ebdb-4476-bae3-08f188f1b544

Conflicts:
	applications/plugins/SofaPython/Binding_BaseContext.cpp

JeremieA committed:
FIX: SofaPython build on linux

Conflicts:
	applications/plugins/SofaPython/CMakeLists.txt
	applications/plugins/SofaPython/PythonEnvironment.cpp

neildavidforrest committed:
FIX: This fixes undefined behaviour and potential crash in Python BaseContext.getObjects() due to PyList accessed out of bounds.

Conflicts:
	applications/plugins/SofaPython/Binding_BaseContext.cpp

neildavidforrest committed:
UPDATE: object and type names are now both optional when calling BaseContext_getObjects(), to match the behavior of the function in the Sofa public repository

Conflicts:
	applications/plugins/SofaPython/Binding_BaseContext.cpp

neildavidforrest committed:
UPDATE: the search direction can now optionally be passed to BaseContext_getObjects() (default is 'Local')

Conflicts:
	applications/plugins/SofaPython/Binding_BaseContext.cpp

neildavidforrest committed:
FIX: solve a memory leak which happens whenever an object is transferred from C++ to Python.

Conflicts:
	applications/plugins/SofaPython/PythonMacros.h

fjourdes committed:
UPDATE: Binding_Data. Expose fullPath read only parameter for BaseData that derive from DataFileName, which as the name suggests, returns the fullPath to the file pointed by that data. Returns None otherwise.

Conflicts:
	applications/plugins/SofaPython/Binding_Data.cpp

Commit:
make the adaptation of the commit 0de46d9abb94 with the new architecture of master, modification of PythonFactory. Creation of Binding_BaseTopologyObject common to PointSetTopologyModifier and TriangleSetTopologyModifier

guparan committed:
Scene testing: Fix warning detection and differentiate errors and crashes.

guparan committed:
FIX: CI should count crashes now (not errors).

guparan committed:
Merge remote-tracking branch 'github/issofa_python' into reset_issofa_integration

guparan committed:
CI: add -DSOFA_BUILD_TUTORIALS=ON to CI default build options.

[ci-ignore] (add this tag to your commits to skip CI builds)

guparan committed:
EDIT: Make test-scenes.sh able to test any folder
[ci-ignore]

guparan committed:
EDIT: CI: Differentiate errors and crashes in Examples testing. Improve logs.
[ci-ignore]

guparan committed:
EDIT: CI log consistency
[ci-ignore]

guparan committed:
CI: Fix scene-tests log indentation
[ci-ignore]

guparan committed:
[CI] FIX: errors and crashes count in summary.
[ci-ignore]

guparan committed:
[Tutorials] FIX file not found on include

guparan committed:
[CI] FIX runSofa_d finding if Debug build on Windows

guparan committed:
[CI] Use CI_OPTIONS instead of job name for plugins-building flag

remibessard committed:
ADD regression tests for several key components in RESET project (scenes from examples folder, references generated in reset_issofa_integration branch at commit 769e6d52c43f)

remibessard committed:
UPDATE: let's say that the reference files are regenerated and that there is no copy of the scene files from the sofa/examples in the regression folders anymore

remibessard committed:
FIX(try): lower the epsilon in order to let the caduceus test pass

guparan committed:
[SofaTest] CLEAN + FIX Windows path length errors

guparan committed:
[SofaGui] FIX Qt dll copy on Windows
Qt dlls are needed by runSofa, even in batch mode (scene tests on CI).
[ci-ignore]

guparan committed:
[SofaKernel] FIX out of bounds access in apply() with TopologicalModifiers scenes

guparan committed:
[CI] EDIT Bigger timeout for scene tests with Debug build

[ci-ignore]

guparan committed:
[SofaKernel] REVERT commit 32b591d7297bf57a298ba813f663555b438ddf2a

[SofaKernel] FIX out of bounds access in apply() with TopologicalModifiers scenes

guparan committed:
[SofaKernel] FIX out of bounds access in apply() with TopologicalModifiers scenes

guparan committed:
Merge remote-tracking branch 'origin/master' into reset_issofa_integration

# Conflicts:
#	applications/plugins/SofaPython/Binding.cpp
#	applications/plugins/SofaPython/Binding_BaseContext.cpp
#	applications/plugins/SofaPython/Binding_Data.cpp
#	applications/plugins/SofaPython/PythonFactory.h
#	scripts/ci/compile.sh

guparan committed:
[SofaPython] FIX merge error

See c64283debc9aac89ba26dc07d76bdbf0e0faff16

matthieu-nesme commented on SofaKernel/modules/SofaBaseMechanics/IdentityMapping.inl:
```diff
@@ -75,6 +75,11 @@ void IdentityMapping<TIn, TOut>::apply(const core::MechanicalParams * /*mparams*
     helper::ReadAccessor< Data<InVecCoord> > in = dIn;
 
+    if(out.size() != in.size()) // true if topological changes
+    {
```
This has nothing to do with SofaPython.
IMPORTANT: NEVER resize just a vector of a mstate. It is an essential rule. You have to resize through BaseState::resize (that resizes every state vectors of the mstate). And then the test can be done only once per step in Apply, and not in every ApplyJ.

JeremieA replied:
This indeed has nothing to do with SofaPython. It was added in cf5a31a5 "FIX out of bounds access in apply() with TopologicalModifiers scenes", I guess related to trying to fix crashes in tests or sample scenes. But I agree this is not the right way to fix this kind of issues.

guparan replied:
Yes that is exactly why we added this change. I will revert it.

matthieu-nesme commented on applications/plugins/SofaPython/Binding_Data.cpp:
```diff
@@ -670,6 +674,26 @@ bool SetDataValuePython(BaseData* data, PyObject* args)
+    BaseData* data = ((PyPtr<BaseData>*)self)->object; // TODO: check dynamic cast
+
+    if (sofa::core::objectmodel::DataFileName* dataFilename = dynamic_cast<sofa::core::objectmodel::DataFileName*>(data))
+    {
```
Rather than perform a dynamic_cast on DataFileName in generic binding_Data.
The clean way is to create a new binding for DataFileName and to add it to the Factory.

guparan replied:
@Ngautier ?

JeremieA replied:
The original commit predates the concept of the `PythonFactory`. So it could indeed be refactored now, by someone who knows how to use this Factory.

matthieu-nesme commented on applications/plugins/SofaPython/Binding_Data.cpp:
```diff
@@ -110,31 +112,34 @@ PyObject *GetDataValuePython(BaseData* data)
     const void* valueVoidPtr = data->getValueVoidPtr();
 
-    if (!typeinfo->Container())
+    if (!typeinfo->Container() || typeinfo->ValidInfo())
```
What is the idea with typeinfo->ValidInfo()? Can you explain please?!

What does happen if the typeinfo is both a container and validinfo?
Is the idea is to remove the default implementation (returns a string) if !typeinfo->ValidInfo()?

guparan replied:
@Ngautier ?

JeremieA replied:
It's from an old change on our end (December 2014). But the original commit only added `if(typeinfo->ValidInfo())` around all the codes using `typeinfo`, `!typeinfo->Container()` was not there. So this is a merge error.

Given that this code is now completely changed in our version (with a redesign of `DataTypeInfo`), and probably conflicts with other changes in the master, I think this commit could simply be removed from this PR.

guparan committed:
[SofaKernel] REVERT commit cf5a31a50f9cb1c6897c7b39c52d0b4034412efb

[SofaKernel] FIX out of bounds access in apply() with TopologicalModifiers scenes (reverted from commit cf5a31a50f9cb1c6897c7b39c52d0b4034412efb)

guparan committed:
[SofaPython] REVERT commit 8ceca2e646f365f36ce66d5f5b6d23f88f771f05

CHANGE : allow to get the value of a data as a string when its type is not handle by the binding

guparan commented:
Reverts done. Tell me if everything is OK :)

matthieu-nesme commented:
- There still is a dubious  `else if (typeinfo->ValidInfo())` in binding_data.
- DataFileName should be handled as its own type.
- how not to push regression test results? so everyone do not have to download it. If I well remember this files are created automatically the first time the test is launched, and then used. Is it not enough?

guparan committed:
[SofaPython] REVERT commit edf11b0dc22cc08263a0b758004a3dc06ca106b6

CHANGE : if the data to be gotten using the bind is a vector of type 'unknown' then the text of the data is returned as a whole
MERGE CHANGE : include was already here, put amelioration : test line 137 on typeinfo->ValidInfo cleaner than test on name

git-svn-id: https://code.insimo.fr/svn/IS/ISSofa/branches/ISSofa@10519 f8a8aa39-ebdb-4476-bae3-08f188f1b544

guparan commented:
The remaining `else if (typeinfo->ValidInfo())` is now reverted.

About the regression test results, we needed a reference (diff base) for each example we wanted to test. We chose to run regression tests on these examples :
```
applications/plugins/SofaTest/SofaTest_test/regression/gravity.scn
examples/Components/collision/FFDDistanceGridCollisionModel.scn
examples/Components/collision/RuleBasedContactManager.scn
examples/Components/constraint/BilateralInteractionConstraint.scn
examples/Components/constraint/FixedConstraint.scn
examples/Components/constraint/FrictionContact.scn
examples/Components/forcefield/BeamFEMForceField.scn
examples/Components/forcefield/FastTriangularBendingSprings.scn
examples/Components/forcefield/TriangularFEMForceFieldOptim.scn
examples/Components/interactionforcefield/InteractionEllipsoidForceField.scn
examples/Components/topology/Mesh2PointTopologicalMapping.scn
examples/Demos/caduceus.scn
```
We can't use regression tests without these reference files.

matthieu-nesme committed:
[SofaPython] binding DataFileName as its own type

matthieu-nesme commented:
I just implemented DataFileName binding.

This PR could be merged as soon as the regression tests are moved somewhere else (the idea is to create a new git containing all regression tests with associated scenes and data).

matthieu-nesme committed:
Merging master

damienmarchal commented:
[ci-build]

guparan committed:
[SofaTest] ADD usage of msg_* api

guparan committed:
[SofaTest] REMOVE regression scenes and references

All regression scenes references have been moved to their own repository:
https://github.com/sofa-framework/regression

damienmarchal commented:
[ci-build]
